MINOR: Optimizer example and docs, deprecate `Expr::name` (#3788)
apache/arrow-datafusion: datafusion/optimizer/README.md

//
// The `datafusion-expr` crate provides a LogicalPlanBuilder
// The `datafusion-sql` crate provides a SQL query planner that can create a LogicalPlan from SQL
// The `datafusion` crate provides a DataFrame API that can create a LogicalPlan
let logical_plan = ...

let mut config = OptimizerConfig::default();
let optimizer = Optimizer::new(&config);
let optimized_plan = optimizer.optimize(&logical_plan, &mut config, observe)?;

fn observe(plan: &LogicalPlan, rule: &dyn OptimizerRule) {
    println!(
        "After applying rule '{}':\n{}",
        rule.name(),
        plan.display_indent()
    )
}
```

## Providing Custom Rules

The optimizer can be created with a custom set of rules.

```rust
let optimizer = Optimizer::with_rules(vec![
    Arc::new(MyRule {})
]);
```

## Writing Optimization Rules

Please refer to the [examples](examples) to learn more about the general approach to writing optimizer rules and
then move onto studying the existing rules.

All rules must implement the `OptimizerRule` trait.

```rust
/// `OptimizerRule` transforms one ['LogicalPlan'] into another which
/// computes the same results, but in a potentially more efficient

> /// way.

Reviewer: Does it make sense to emphasize that in the event of there are no transformations to apply (for the given input), the optimizer could (and should) return the input as is (e.g. instead of failing with an exception)?

```
/// `OptimizerRule` can transform one ['LogicalPlan'] into another which
/// computes the same results, but in a potentially more efficient way. If
/// there are no suitable transformations for the input plan, the optimizer can
/// simply return it as is.

Author: That's a very good point. I will add this.

  under the License.
-->

# DataFusion Query Optimizer

[DataFusion](df) is an extensible query execution framework, written in Rust, that uses Apache Arrow as its in-memory
format.

DataFusion has modular design, allowing individual crates to be re-used in other projects.

> This crate is a submodule of DataFusion that provides a query optimizer for logical plans.

Reviewer 2: ```suggestion
This crate is a submodule of DataFusion that provides a query optimizer for logical plans, and
contains an extensive set of OptimizerRules that may rewrite the plan and/or its expressions so
they execute more quickly while still computing the same result.
```

individual operators or expressions.

Sometimes there is an initial pass that visits the plan and builds state that is used in a second pass that performs
the actual optimization. This approach is used in projection push down and filter push down.

### Expression Naming

Every expression in DataFusion has a name, which is used as the column name. For example, in this example the output
contains a single column with the name `"COUNT(aggregate_test_100.c9)"`:

```text
❯ select count(c9) from aggregate_test_100;
+------------------------------+
| COUNT(aggregate_test_100.c9) |
+------------------------------+
| 100                          |
+------------------------------+
```

These names are used to refer to the columns in both subqueries as well as internally from one stage of the LogicalPlan
to another. For example:

```text
❯ select "COUNT(aggregate_test_100.c9)" + 1 from (select count(c9) from aggregate_test_100) as sq;
+--------------------------------------------+
| sq.COUNT(aggregate_test_100.c9) + Int64(1) |
+--------------------------------------------+
| 101                                        |
+--------------------------------------------+
```

### Implication

> 
> DataFusion contains an extensive set of OptimizerRules that may rewrite the plan and/or its expressions so they execute
> more quickly.

Reviewer 2: ```suggestion
```

Suggest moving this up

consuming `self` producing a new expression.

```rust
let mut expr_rewriter = MyExprRewriter {};
let expr = expr.rewrite(&mut expr_rewriter)?;
```

Here is an example implementation which will rewrite `expr BETWEEN a AND b` as `expr >= a AND expr <= b`. Note that the
implementation does not need to perform any recursion since this is handled by the `rewrite` method.

```rust
struct MyExprRewriter {}

impl ExprRewriter for MyExprRewriter {
    fn mutate(&mut self, expr: Expr) -> Result<Expr> {
        match expr {
            Expr::Between {
                negated,
                expr,
                low,
                high,
            } => {
                let expr: Expr = expr.as_ref().clone();
                let low: Expr = low.as_ref().clone();
                let high: Expr = high.as_ref().clone();
                if negated {
                    Ok(expr.clone().lt(low).or(expr.clone().gt(high)))
                } else {
                    Ok(expr.clone().gt_eq(low).and(expr.clone().lt_eq(high)))
                }
            }
            _ => Ok(expr.clone()),
        }
    }
}
```

### optimize_children

> It is quite typical for a rule to be applied recursively to all operators within a query plan. Rather than duplicate
> that logic in each rule, an `optimize_children` method is provided. This recursively invokes the `optimize` method on
> the plan's children and then returns a node of the same type.

Reviewer 2: ```suggestion
Typically a rule is applied recursively to all operators within a query plan. Rather than duplicate
that logic in each rule, an `optimize_children` method is provided. This recursively invokes the `optimize` method on
the plan's children and then returns a node of the same type.
```

```rust
fn optimize(
    &self,
    plan: &LogicalPlan,
    _config: &mut OptimizerConfig,
) -> Result<LogicalPlan> {
    // recurse down and optimize children first
    let plan = utils::optimize_children(self, plan, _config)?;

    ...
}
```

### Writing Tests

There should be unit tests in the same file as the new rule that test the effect of the rule being applied to a plan
in isolation (without any other rule being applied).

There should also be a test in `integration-tests.rs` that tests the rule as part of the overall optimization process.

### Debugging

> The `EXPLAIN VERBOSE` command can be used to show the effect of each optimization rule on a query.

Reviewer 2: ```suggestion
The `EXPLAIN VERBOSE` command can be used to show the effect of each optimization rule on a query.

In the following example, the `type_coercion` and `simplify_expressions` passes have simplified the plan so that it  returns the constant `"3.2"` rather than doing a computation at execution time. 
```
